optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/DroolsPatternBuilder.java

```diff
@@ -0,0 +1,126 @@
+ * If the pattern inside Stream 1 was shared by Stream 2 and Stream 3, filters from Stream 2 would also apply to
+ * Stream 3, and vice versa.
+ * Therefore, Stream 2 and Stream 3 each get their own immutable copy of Stream 1's {@link DroolsPatternBuilder}, one
+ * which only applies their individual filters.
+ * @param <T> generic type of the pattern to be eventually created by this builder
+ */
+public final class DroolsPatternBuilder<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DroolsPatternBuilder.class);
+    private final Variable<T> baseVariable;
+    private final Function<PatternDSL.PatternDef<T>, PatternDSL.PatternDef<T>> builder;
+
+    /**
+     * @param baseVariable will be used during the {@link PatternDSL#pattern(Variable)} call within {@link #build()} or
+     * {@link #build(DeclarationSource)}
+     */
+    public DroolsPatternBuilder(Variable<T> baseVariable) {
+        this.baseVariable = baseVariable;
+        this.builder = p -> {
+            LOGGER.trace("Starting pattern for {}.", baseVariable);
+            return p;
+        };
+    }
+
+    private DroolsPatternBuilder(Variable<T> baseVariable,
+            Function<PatternDSL.PatternDef<T>, PatternDSL.PatternDef<T>> builder) {
+        this.baseVariable = baseVariable;
+        this.builder = builder;
+    }
+
+    /**
+     * Mutate the existing {@link PatternDSL.PatternDef}, adding a new operation, such as a new filter or a new variable
+     * binding.
+     * @param expander operation to mutate the stream
+     * @return copy of the {@link DroolsPatternBuilder} with the new expanding operation included.
+     */
+    public DroolsPatternBuilder<T> expand(UnaryOperator<PatternDSL.PatternDef<T>> expander) {
+        Function<PatternDSL.PatternDef<T>, PatternDSL.PatternDef<T>> newBuilder = builder.andThen(expander);
+        return new DroolsPatternBuilder<>(baseVariable, newBuilder);
+    }
+
+    /**
+     * Mutate the existing {@link PatternDSL.PatternDef}, adding a new operation, such as a new filter or a new variable
+     * binding.
+     * @param description descriptive name for the operation, to be printed in trace logging
+     * @param expander operation to mutate the stream
+     * @return copy of the {@link DroolsPatternBuilder} with the new expanding operation included.
+     */
+    public DroolsPatternBuilder<T> expand(String description, UnaryOperator<PatternDSL.PatternDef<T>> expander) {
+        UnaryOperator<PatternDSL.PatternDef<T>> expanderWithLogging = p -> {
+            LOGGER.trace("Expanding {} pattern with '{}'.", baseVariable, description);
+            return expander.apply(p);
+        };
+        return expand(expanderWithLogging);
+    }
+
+    private PatternDSL.PatternDef<T> build(PatternDSL.PatternDef<T> basePattern) {
+        PatternDSL.PatternDef<T> result = builder.apply(basePattern);
+        LOGGER.trace("Pattern for {} finished: {}.", baseVariable, result);
```

Comment:
We encourage users to turn on "TRACE" logging to see the actual moves that optaplanner is taking.
Therefore, even the trace log must be _readable_ and contain useful information only.
One trick that often works well, is to ident trace logging with 2 spaces, so it's easier to see "solving started" and "step ended" log lines etc.

If these trace loggings are purely here for our debugging purposes, I recommend to use "debugger log messages" instead.
I think it would be useful to see which drools constraints are build from the CS, in trace logging, at the end of creating a constraint. I am not sure it's useful to see both "start" and "end" messages - I am not sure if those are worth their weight.

Reply:
Wrt. indentation - can do.

Wrt."debugger log messages" - I'm probably not familiar with that pattern. Can you point me someplace?